File: abc215/d.cpp
#include <set>
#include <vector>
#include <iostream>
using namespace std;


void prime_factorize(long long n, set<long long>& primes){
    primes.insert(1);
    while(n % 2 == 0){
        primes.insert(2);
        n /= 2;
    }
    long long f = 3;
    while (f*f <= n){
        if (n % f == 0){
            primes.insert(f);
            n /= f;
        }else{
            f += 2;
        }
    }
    if (n != 1) primes.insert(n);
}
/*
long long GCD(unsigned long long a, unsigned long long b) {
    if (b == 0) return a;
    else return GCD(b, a % b);
}
*/
int main(){
    long long N, M;
    cin >> N >> M;
    std::vector<long long> A(N);
    set<long long> primes;
    for (long long i=0; i<N; i++) cin >> A.at(i);
    for (long long i=0; i<N; i++){
        prime_factorize(A[i], primes);
    }
    //cout << "primes: " << primes.size() << endl;
    // 以下の方法では、lが超巨大になるので、64bitでも計算できない。
    /*unsigned long long l=1;
    for (auto x : primes){
        l *= x;
    }
    cout << "l: " << l << endl;
    */
    vector<long long> ans;
    for (long long k = 1; k <= M; k++){
        bool valid = true;
        for (auto p: primes){
            if (p != 1 && k%p == 0){
                valid = false;
                break;
            }
        }
        if (valid) ans.push_back(k);
    }
    cout << ans.size() << endl;
    for (long long i = 0; i < ans.size(); i++){
        cout << ans[i] << endl;
    }

    return 0;
}

File: abc215/d.py
# Aiに出てくる素因数の積lを計算し、各kでgcdが1かどうかを調べれば、
# O(N*max(A_i)^(1/2)+M)で間に合いそうだが、
# lが非常に大きくなるのでTLEする。
# C++などに書き換えた場合、lが64bitに収まらないのでWAになる。
# よく考えると、Aに含まれる素数が5桁だとするとlは10^(5*10^5)程度になり、クソでかい。
# そこで、Aに含まれる素数は10000個程度なので、kを割りきれるかどうかを調べれば間に合う。
# x以下の素数の個数をπ(x)と書くと、計算量はO(N*max(A_i)^(1/2)+Mπ(max(A_i))
# こう書くとMπ(max(A_i)の部分がギリギリに見えるが、C++はとても速いので間に合う
# (あと割りきれるかどうかを調べるだけならはやい)。
# 解説にあるように、エラストテネスの篩を利用して既に確認した因数を持つ候補を消していく
# 方がスマートでしょう…
import math
#from collections import defaultdict

N, M = [int(s) for s in input().split()]
A = [int(s) for s in input().split()]
primes = set()
def prime_factorize(n):
    primes.add(1)
    while n % 2 == 0:
        primes.add(2)
        n //= 2
    f = 3
    while f * f <= n:
        if n % f == 0:
            primes.add(f)
            n //= f
        else:
            f += 2
    if n != 1:
        primes.add(n)
    return a


for a in A:
    prime_factorize(a)

l = primes.pop()
while len(primes):
    l *= primes.pop()
ans = []
for k in range(1, M+1):
    if math.gcd(l, k) == 1:
        ans.append(k)
print(len(ans))
for a in ans:
    print(a)   

'''
import functools
def my_lcm_base(x, y):
    return (x * y) // math.gcd(x, y)

def my_lcm(*integers):
    return functools.reduce(my_lcm_base, integers)
lcm = my_lcm(*A)
#print(lcm)
ans = []
for k in range(1, M+1):
    if math.gcd(lcm, k) == 1:
        ans.append(k)
print(len(ans))
for a in ans:
    print(a)        
'''


